src/coreclr/tests/src/JIT/opt/InstructionCombining/ArrayLengthArithmetic.cs

```diff
@@ -0,0 +1,143 @@
+        for (int arrayLength = 0; arrayLength < 100; arrayLength++)
+        {
+            var array = new int[arrayLength];
+
+            if (arrayLength == 0)
+            {
+                Expect<DivideByZeroException>(() => ArrayLengthDiv_cns0(array));
+                Expect<DivideByZeroException>(() => ArrayLengthDiv_var0(array));
+                Expect<DivideByZeroException>(() => ArrayLengthMod_cns0(array));
+                Expect<DivideByZeroException>(() => ArrayLengthMod_var0(array));
+            }
+
+            // Array.Length / cns == Array.Length / ToVar(cns)
+            CompareResults(ArrayLengthDiv_cns1(array), ArrayLengthDiv_var1(array));
+            CompareResults(ArrayLengthDiv_cns2(array), ArrayLengthDiv_var2(array));
+            CompareResults(ArrayLengthDiv_cns3(array), ArrayLengthDiv_var3(array));
+            CompareResults(ArrayLengthDiv_cns4(array), ArrayLengthDiv_var4(array));
+            CompareResults(ArrayLengthDiv_cns5(array), ArrayLengthDiv_var5(array));
+            CompareResults(ArrayLengthDiv_cns8(array), ArrayLengthDiv_var8(array));
+            CompareResults(ArrayLengthDiv_cns10(array), ArrayLengthDiv_var10(array));
+            CompareResults(ArrayLengthDiv_cnsMaxValuen1(array), ArrayLengthDiv_varMaxValuen1(array));
+            CompareResults(ArrayLengthDiv_cnsMaxValue(array), ArrayLengthDiv_varMaxValue(array));
+            CompareResults(ArrayLengthDiv_cnsn1(array), ArrayLengthDiv_varn1(array));
+            CompareResults(ArrayLengthDiv_cnsn2(array), ArrayLengthDiv_varn2(array));
+            CompareResults(ArrayLengthDiv_cnsMinValue(array), ArrayLengthDiv_varMinValue(array));
+
+            // Array.Length % cns == Array.Length % ToVar(cns)
+            CompareResults(ArrayLengthMod_cns1(array), ArrayLengthMod_var1(array));
+            CompareResults(ArrayLengthMod_cns2(array), ArrayLengthMod_var2(array));
+            CompareResults(ArrayLengthMod_cns3(array), ArrayLengthMod_var3(array));
+            CompareResults(ArrayLengthMod_cns4(array), ArrayLengthMod_var4(array));
+            CompareResults(ArrayLengthMod_cns5(array), ArrayLengthMod_var5(array));
+            CompareResults(ArrayLengthMod_cns8(array), ArrayLengthMod_var8(array));
+            CompareResults(ArrayLengthMod_cns10(array), ArrayLengthMod_var10(array));
+            CompareResults(ArrayLengthMod_cnsMaxValuen1(array), ArrayLengthMod_varMaxValuen1(array));
+            CompareResults(ArrayLengthMod_cnsMaxValue(array), ArrayLengthMod_varMaxValue(array));
+            CompareResults(ArrayLengthMod_cnsn1(array), ArrayLengthMod_varn1(array));
+            CompareResults(ArrayLengthMod_cnsn2(array), ArrayLengthMod_varn2(array));
+            CompareResults(ArrayLengthMod_cnsMinValue(array), ArrayLengthMod_varMinValue(array));
+        }
+
+        return returnCode;
+    }
+
+    // Array.Length / cns
+    [MethodImpl(MethodImplOptions.NoInlining)] private static int ArrayLengthDiv_cns0(int[] array) => array.Length / 0;
```

Comment:
>ArrayLengthDiv_cns0 [](start = 66, length = 19)

should we add cases where the operation could become CSE candidate?
```
 .. = array.Length / 3;
...
.. = array.Length / 3;
```

Reply:
I do not see any potential issues here, because we don't change `tree->gtType`, so I can't construct a test where this transformation can lead to add new CSE candidates or reject an existing.
It would be nice to improve the test even more, but I would not block the PR because of that.